hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/ByteArrayObjectDataOutput.java

```diff
@@ -249,15 +246,19 @@ public void writeShort(int position, int v, ByteOrder byteOrder) throws IOExcept
 
     @Override
     public void writeUTF(final String str) throws IOException {
-        if (utfBuffer == null) {
-            utfBuffer = new byte[UTF_BUFFER_SIZE];
+        int len = (str != null) ? str.length() : NULL_ARRAY;
+        writeInt(len);
+        if (len > 0) {
+            ensureAvailable(len * 3);
+            for (int i = 0; i < len; i++) {
+                pos += Bits.writeUtf8Char(buffer, pos, str.charAt(i));
```

Comment:
I have a suspicion this might not be right. If the char is part of a surrogate pair then you need both the low surrogate and the surrogate to convert it into a utf-8 character.